### DISCUSSION TITLE
Edge regression on heterogeneous graphs

### DISCUSSION THREAD

**Discussion by RENEK-bool** [pyg-team/pytorch_geometric#9037] [Q&A] [answered]
Recently I wanted to use GNN to solve the problem, and I built several heterogeneous graphs for my problem, each heterogeneous graph has only one type of node and two types of edges. For the sake of illustration, I will use A and B to represent the two types of edges. There are no features on edge A there are no labels, there are no features on nodes, there are only features and labels on edge B, and then I want to be able to predict the value of edge B by GNN. Since I don't know much about GNN, I feel as if there is no model that can implement this idea, and it seems that GNN generally handles graphs that need to have node features.

Here is my definition of a heterogeneous graph
```
data = HeteroData()

data['pin', 'cell_arc', 'pin'].edge_index = torch.tensor(cell_edge_index, dtype=torch.long).t().contiguous()
data['pin', 'net_arc', 'pin'].edge_index = torch.tensor(net_edge_index, dtype=torch.long).t().contiguous()

data['pin', 'net_arc', 'pin'].edge_attr = torch.tensor(net_arc_feature, dtype=torch.float)
```

**Comment by RENEK-bool** [comment]
I have another question: For a heterogeneous graph, which has two types of nodes, one type of edge, and only one type of node has features and labels, can there be a model to complete the node regression task (I noticed that pytorch geometric has the function of automatically converting to heterogeneous GNN model)?

**Comment by rusty1s** [comment, marked as answer]
You can well use GNNs for this. The main question is how you wanna deal with "empty" features:
* You can use a shallow embedding to learn a node feature representation (via `torch.nn.Embedding`)
* You can just interpret missing features as `torch.ones(num_nodes, 1)`

**Reply by RENEK-bool** [reply]
Thank you for your answer, do you mean that both of my ideas need to deal with empty features, that is, node features are necessary?One more question I have is do edge features pass messages? Or is it only the node features that are used for message passing?

**Reply by rusty1s** [reply]
If you want to use GNNs, then you will need a set of input node features to begin with. GNNs can also integrate edge features into its message passsing scheme (see our GNN cheatsheet in the documentation for an overview of supported layers), but they will always require node features.

**Reply by RENEK-bool** [reply]
OK, thank you.
